runtime/pkg/rduckdb/catalog.go

```diff
@@ -0,0 +1,253 @@
+/*
+Example init logic:
+- Sync remote files with the local cache
+- Create a catalog
+- Traverse the local files and call addTableVersion for table
+Example write logic:
+- Call addTableVersion after adding a new table version
+- Call removeTable when deleting a table
+Example read logic:
+- Call acquireSnapshot when starting a read
+- If it doesn't already exist, create a schema for the snapshot ID with views for all the table version in the snapshot
+- Call releaseSnapshot when done reading the snapshot
+Example removeFunc logic:
+- Detach the version
+- Remove the version file
+- If there are no files left in it, remove the table folder
+*/
+package rduckdb
+
+import (
+	"context"
+	"fmt"
+
+	"golang.org/x/sync/semaphore"
+)
+
+// Represents one table and its versions currently present in the local cache.
+type table struct {
+	name                   string
+	deleted                bool
+	currentVersion         string
+	versionReferenceCounts map[string]int
+	versionMeta            map[string]*tableMeta
+}
+
+// Represents a snapshot of table versions.
+// The table versions referenced by the snapshot are guaranteed to exist for as long as the snapshot is acquired.
+type snapshot struct {
+	id             int
+	referenceCount int
+	tables         []*tableMeta
+	// if snapshot is ready to be served then ready will be marked true
+	ready bool
+}
+
+// Represents a catalog of available table versions.
+// It is thread-safe and supports acquiring a snapshot of table versions which will not be mutated or removed for as long as the snapshot is held.
+type catalog struct {
+	sem               *semaphore.Weighted
+	tables            map[string]*table
+	snapshots         map[int]*snapshot
+	currentSnapshotID int
+
+	removeVersionFunc  func(context.Context, string, string) error
+	removeSnapshotFunc func(context.Context, int) error
+}
+
+// newCatalog creates a new catalog.
+// The removeSnapshotFunc func will be called exactly once for each snapshot ID when it is no longer the current snapshot and is no longer held by any readers.
+// The removeVersionFunc func will be called exactly once for each table version when it is no longer the current version and is no longer used by any active snapshots.
+func newCatalog(removeVersionFunc func(context.Context, string, string) error, removeSnapshotFunc func(context.Context, int) error) *catalog {
+	return &catalog{
+		sem:                semaphore.NewWeighted(1),
+		tables:             make(map[string]*table),
+		snapshots:          make(map[int]*snapshot),
+		removeVersionFunc:  removeVersionFunc,
+		removeSnapshotFunc: removeSnapshotFunc,
+	}
+}
+
+func (c *catalog) tableMeta(ctx context.Context, name string) (*tableMeta, error) {
```

Comment:
Yes I had used mutex to pass to `removeVersionFunc` and `removeSnapshotFunc`. Because I already had mutex I thought I might as well use a semaphore to cancel early.
I thought may be not cleaning entirely is fine since we frequently rename and drop tables so it is fine to leave background cleanup once in a while but background goroutine can also be considered. 
Modified the code so that most catalog functions do not accept ctx and do not return error as well. Slightly easier to integrate with db.